keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+
+Here is an example of how to define a pluralized MY_POD_IPS environmental variable in a pod definition yaml file:
+```
+  - name: MY_POD_IPS
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs
+  - name: MY_POD_IPS_PROPERTIES_0
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs[0].properties
+```
+
+This definition will cause an environmental variable setting in the pod similar to the following:
+```
+MY_POD_IPS=fd00:10:20:0:3::3,10.20.3.3
+MY_POD_IPS_PROPERTIES_<IP_index>=x:y #We assume that colon is not going to used in map values.
+```
+
+### Kubeadm Support
+
+Dual-stack support will need to be added to kubeadm both for dual-stack development purposes, and for use in dual-stack continuous integration tests.
+
+- The Kubeadm config options and config file will support dual stack options for apiserver-advertise-address, and podSubnet.
+
+#### Kubeadm Configuration Options
+
+The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:
+```
+    api:
+      advertiseAddress: "fd00:90::2,10.90.0.2" [Multiple IP CIDRs, comma separated list of CIDRs]
+    networking:
+      podSubnet: "fd00:10:20::/72,10.20.0.0/16" [Multiple IP CIDRs, comma separated list of CIDRs]
+```
+
+#### Kubeadm-Generated Manifests
+
+Kubeadm will need to generate dual-stack CIDRs for the --service-cluster-ip-range command line argument in kube-apiserver.yaml:
+```
+    spec:
+      containers:
+      - command:
+        - kube-apiserver
+        - --service-cluster-ip-range=fd00:1234::/110,10.96.0.0/12
+```
+
+Kubeadm will also need to generate dual-stack CIDRs for the --cluster-cidr argument in kube-apiserver.yaml:
+```
+    spec:
+      containers:
+      - command:
+        - kube-controller-manager
+        - --cluster-cidr=fd00:10:20::/72,10.20.0.0/16
+```
+
+### vendor/github.com/spf13/pflag
+This dual-stack proposal will introduce a new IPNetSlice object to spf13.pflag to allow parsing of comma separated CIDRs. Refer to [https://github.com/spf13/pflag/pull/170](https://github.com/spf13/pflag/pull/170)
+
+### End-to-End Test Support
+End-to-End tests will be updated for Dual Stack. The Dual Stack E2E tests will use deployment scripts from the kubernetes-sigs/kubeadm-dind-cluster github repo to set up a containerized, multi-node Kubernetes cluster that is running in a Prow container (Docker-in-Docker-in-Docker, or DinDinD configuration), similar to the IPv6-only E2E tests (see [test-infra PR # 7529](https://github.com/kubernetes/test-infra/pull/7529)). The DinDinD cluster will be updated to support dual-stack.
```

Comment:
I'm working in adding IPv6 support to `kind` and I plan to add jobs to test IPv6 and DualStack scenarios later with `kind`

https://github.com/kubernetes-sigs/kind/pull/348

Reply:
Will add this to the KEP